### PR TITLE
fix: Change Argo Workflows executor

### PR DESCRIPTION
**Pull request by sibuthomasmathew**
The container runtime executor for workflows defaults to docker.
Since Civo kubernetes is based on k3s, containerd is used instead
of docker. So jobs submitted to workflows fail unless the container
rutime executor is updated to pns(process namespace sharing)

Signed-off-by: Sibu Thomas Mathew <sibuthomasmathew@gmail.com>

Thank you for wanting to submit a Pull Request to the Civo Kubernetes Marketplace repository!

If your pull request concerns an existing Marketplace application, please make sure you have:

* [x] Notified the Maintainer of the application in this pull request so that they are aware of your proposal.
* [x] Outlined the changes you are proposing, and the reasons these are required (e.g. stability, compatibility with new versions of Kubernetes implementations, etc).
* [x] Tested that the application works with the proposed updates applied (including a screenshot)

## Output before the change

```bash
$ kubectl get nodes
NAME                                                     STATUS   ROLES    AGE   VERSION
k3s-accure-workflow-cluster-776d-63c6ae-node-pool-1b25   Ready    <none>   10m   v1.20.2+k3s1
k3s-accure-workflow-cluster-776d-63c6ae-node-pool-7a04   Ready    <none>   12m   v1.20.2+k3s1
$ 
$ 
$ kubectl -n argo describe cm workflow-controller-configmap
Name:         workflow-controller-configmap
Namespace:    argo
Labels:       <none>
Annotations:  <none>

Data
====

BinaryData
====

Events:  <none>
$ 
$ 
$ kubectl get pods
NAME                READY   STATUS              RESTARTS   AGE
hello-world-zkqsh   0/2     ContainerCreating   0          6m41s
$ 
$ 
$ kubectl describe pod hello-world-zkqsh | tail
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason       Age                  From               Message
  ----     ------       ----                 ----               -------
  Normal   Scheduled    7m8s                 default-scheduler  Successfully assigned default/hello-world-zkqsh to k3s-accure-workflow-cluster-776d-63c6ae-node-pool-1b25
  Warning  FailedMount  56s (x11 over 7m8s)  kubelet            MountVolume.SetUp failed for volume "docker-sock" : hostPath type check failed: /var/run/docker.sock is not a socket file
  Warning  FailedMount  29s (x3 over 5m5s)   kubelet            Unable to attach or mount volumes: unmounted volumes=[docker-sock], unattached volumes=[podmetadata docker-sock default-token-gwjzd]: timed out waiting for the condition
$ 
```

## After the change

```bash
$ kubectl get nodes
NAME                                                     STATUS   ROLES    AGE   VERSION
k3s-accure-workflow-cluster-776d-63c6ae-node-pool-1b25   Ready    <none>   16m   v1.20.2+k3s1
k3s-accure-workflow-cluster-776d-63c6ae-node-pool-7a04   Ready    <none>   17m   v1.20.2+k3s1
$ 
$ 
$ kubectl -n argo describe cm workflow-controller-configmap
Name:         workflow-controller-configmap
Namespace:    argo
Labels:       <none>
Annotations:  <none>

Data
====
config:
----
containerRuntimeExecutor: pns


BinaryData
====

Events:  <none>
$
$  
$ kubectl  get pods -w
NAME                READY   STATUS              RESTARTS   AGE
hello-world-lq9vs   0/2     ContainerCreating   0          11s
hello-world-lq9vs   1/2     NotReady            0          25s
hello-world-lq9vs   0/2     Completed           0          51s
^C$ 
$ 
$ 
$ kubectl describe pod hello-world-lq9vs | tail
  ----    ------     ----  ----               -------
  Normal  Scheduled  72s   default-scheduler  Successfully assigned default/hello-world-lq9vs to k3s-accure-workflow-cluster-776d-63c6ae-node-pool-1b25
  Normal  Pulling    72s   kubelet            Pulling image "argoproj/argoexec:v3.0.3"
  Normal  Pulled     61s   kubelet            Successfully pulled image "argoproj/argoexec:v3.0.3" in 11.544769944s
  Normal  Created    60s   kubelet            Created container wait
  Normal  Started    60s   kubelet            Started container wait
  Normal  Pulling    60s   kubelet            Pulling image "docker/whalesay:latest"
  Normal  Pulled     49s   kubelet            Successfully pulled image "docker/whalesay:latest" in 10.942951694s
  Normal  Created    49s   kubelet            Created container main
  Normal  Started    49s   kubelet            Started container main
$ 
```

### CODE REVIEW

**Comment by sibuthomasmathew** [comment]
Hi @protosam,

Kindly review this PR.

**Comment by protosam** [comment]
lgtm

I assumed Argo just used k8s pods so didn't bother to test more than "did it install".

File: argo-workflows/Makefile
Comment (saiyam1814) [inline argo-workflows/Makefile:8]: Shouldn't the path for --from-file be  `https://raw.githubusercontent.com/argoproj/argo-workflows/config` ?

**Review by saiyam1814 — CHANGES_REQUESTED** [review]

File: argo-workflows/Makefile
Comment (sibuthomasmathew) [reply argo-workflows/Makefile:8]: Hi @saiyam1814 

I'm not sure how that will work because there currently isn't any key-value pair config file available in the official Github repo of Argo workflows. That's why the file `config` with custom configuration was committed so that it can be used to generate `app.yaml` using the `make` command.

Comment (saiyam1814) [reply argo-workflows/Makefile:8]: I need to see how `--from-file=config` will work on the boxes.

Comment (sibuthomasmathew) [reply argo-workflows/Makefile:8]: Without `--from-file=config`, here's how the configmap will look like:
```bash
$ kubectl -n argo describe cm workflow-controller-configmap
Name:         workflow-controller-configmap
Namespace:    argo
Labels:       <none>
Annotations:  <none>

Data
====

BinaryData
====

Events:  <none>
$ 
```

After passing the `config`, here's what the configmap looks like:
```bash
$ kubectl -n argo describe cm workflow-controller-configmap
Name:         workflow-controller-configmap
Namespace:    argo
Labels:       <none>
Annotations:  <none>

Data
====
config:
----
containerRuntimeExecutor: pns


BinaryData
====

Events:  <none>
$
```

The impact on jobs(pods) without and with this configuration change has been shared in the first comment.

If I've failed to answer your query then apologies, I may have not totally understood your question.

Comment (sibuthomasmathew) [reply argo-workflows/Makefile:8]: Btw, `config` is a file that was introduced as part of this commit

Comment (sibuthomasmathew) [reply argo-workflows/Makefile:8]: Looping in the maintainer @protosam for guidance.

Comment (saiyam1814) [reply argo-workflows/Makefile:8]: @protosam what are your thoughts on this PR?

Comment (protosam) [reply argo-workflows/Makefile:8]: Will followup on this over this coming weekend.

Comment (protosam) [reply argo-workflows/Makefile:8]: Using `--from-file=config` is fine. The Makefile is to pre-generate `app.yaml` before commit. The working directory for Makefiles will always be the directory the Makefile is in.

There is a problem though. If you do `kubectl create -f app.yaml` against the pre-generated `app.yaml`, you'll get this error.
```
Error from server (AlreadyExists): error when creating "app.yaml": configmaps "workflow-controller-configmap" already exists
```

The problem is that the install manifest already has a `ConfigMap/workflow-controller-configmap` object. This needs to reduced to a single configmap object.

If you use `kubectl apply`, this isn't an issue, but I don't like that there's two entries for the same ConfigMap.

My only recommendation for change is to introduce using `jq` into the Makefile to filter out the ConfigMap provided from the default manifest. The one that the workflows manifest provides is empty anyways.

```Makefile
build: app.yaml

app.yaml:
	@kubectl create --dry-run=client ns -o yaml argo >> app.yaml
	@echo --- >> app.yaml
	@kubectl -n argo create configmap workflow-controller-configmap --dry-run=client -o yaml --from-file=config >> app.yaml
	@echo --- >> app.yaml
	@kubectl apply --dry-run=client -n argo -o json -f https://raw.githubusercontent.com/argoproj/argo-workflows/stable/manifests/install.yaml | \
		jq 'del(.items[] | \
		select(.metadata.name == "workflow-controller-configmap"))' | \
		kubectl apply --dry-run=client -n argo -o yaml -f - >> app.yaml
	@sed -i -e '/^version:/d' manifest.yaml
	@grep -m1 'image: .*/workflow-controller:' app.yaml | cut -d: -f3 | sed 's/\(.*\)/version: "\1"/' >> manifest.yaml

.PHONY: clean
clean:
	@rm app.yaml
```

Comment (sibuthomasmathew) [reply argo-workflows/Makefile:6]: Hi @protosam,

Thank you for pointing out the problem. As per your suggestion, I've made the correction. 

Kindly review.

Comment (protosam) [reply argo-workflows/Makefile:6]: @saiyam1814 this should be good now. 

This is the workflows documentation for what's being configured as a default for workflows: https://argoproj.github.io/argo-workflows/workflow-executors/#process-namespace-sharing-pns

**Review by saiyam1814 — APPROVED** [review]
LGTM, can you test once it's merged ?

**Pull request merged**
